### PR TITLE
Disallow compactions if there isn't enough free space

### PR DESCRIPTION
**Pull request by amytai**
This diff handles cases where compaction causes an ENOSPC error.
This does not handle corner cases where another background job is started while compaction is running, and the other background job triggers ENOSPC, although we do allow the user to provision for these background jobs with SstFileManager::SetCompactionBufferSize.
It also does not handle the case where compaction has finished and some other background job independently triggers ENOSPC.

Usage: Functionality is inside SstFileManager. In particular, users should set SstFileManager::SetMaxAllowedSpaceUsage, which is the reference highwatermark for determining whether to cancel compactions.

### CODE REVIEW

File: db/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -1541,27 +1542,49 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
-        // compaction, we might be able to execute it in parallel, so we add it
-        // to the queue and schedule a new thread.
-        if (cfd->NeedsCompaction()) {
-          // Yes, we need more compactions!
+        uint64_t size_added_by_compaction = 0;
+        // First check if we even have the space to do the compaction
+        for (size_t i = 0; i < c->num_input_levels(); i++) {
+          for (size_t j = 0; j < c->num_input_files(i); j++) {
+            FileMetaData *filemeta = c->input(i, j);
+            size_added_by_compaction += filemeta->compensated_file_size;
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: probably just use non-compensated file size since the compensation increases the score for files with a lot of deletions, but doesn't represent real space usage

**Comment by amytai** [comment]
It looks like Travis is failing because in one compile it cannot find statvfs: fatal error: "sys/statvfs.h: No such file or directory", is there a way around this? This is a library function.

File: db/db_impl_compaction_flush.cc [hunk 2]
```diff
@@ -1541,27 +1542,49 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
+          }
+        }
+        // Now figure out how much free space is in the file filesystem
+        struct statvfs statbuf;
+        statvfs(dbname_.c_str(), &statbuf);
+        uint64_t freespace = statbuf.f_bsize * statbuf.f_bfree;
+        if (freespace <= size_added_by_compaction) {
+          // Then don't do the compaction
+          c->ReleaseCompactionFiles(status);
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: just curious, does it spam the logs as long as space is not enough? My understanding is this'll loop quickly if there's nothing else in the compaction queue.

File: db/db_impl_compaction_flush.cc [hunk 3]
```diff
@@ -1541,27 +1542,49 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
+        // First check if we even have the space to do the compaction
+        for (size_t i = 0; i < c->num_input_levels(); i++) {
+          for (size_t j = 0; j < c->num_input_files(i); j++) {
+            FileMetaData *filemeta = c->input(i, j);
+            size_added_by_compaction += filemeta->compensated_file_size;
+          }
+        }
+        // Now figure out how much free space is in the file filesystem
+        struct statvfs statbuf;
+        statvfs(dbname_.c_str(), &statbuf);
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: there's a couple issues we need to think more about:

1. this function doesn't exist in Windows, which looks like the cause of travis failing. Maybe make it a function on Env, make it return `Status::NotSupported` by default, and override with a call to `statvfs` in PosixEnv.
2. `dbname_` isn't always the directory compactions will run in. There's also `db_paths` which can spread SST files across multiple directories. I'm ok with just skip this logic if people are setting multiple directories, which is rare anyways.

File: db/db_impl_compaction_flush.cc (hunk 2 above)
Comment (amytai) [reply db/db_impl_compaction_flush.cc]: Right, I forgot to put a sleep here -- is that how we should handle it?

Comment (amytai) [reply db/db_impl_compaction_flush.cc]: I put a temporary sleep there, but we might want to revisit. In particular, in every test I run right now, the main problem happening isn't that compaction is too large, it's the following: Suppose max filesystem size is 10GB. compaction gets triggered, the temporary filesystem size goes up to something like 9.5GB. Compaction doesn't get a chance to clean up, and a logging thread needs to append to a log file with > 0.5GB of data (numbers are relative).
So I wonder if we should actually modify the compaction conditional to cancel compactions if compaction_size + some_buffer_room >= freespace. Where some_buffer_room is a reasonable size for log and other file appends that will be written during the duration of a compaction. For example, if all compactions are approx 3 secs, what is the amount of metadata we expect to be written those 3 secs? That amount should be some_buffer_room.

File: db/db_impl_compaction_flush.cc (hunk 3 above)
Comment (ajkr) [reply db/db_impl_compaction_flush.cc]: For 2., the compaction affects the directory to which it outputs files, which may not be `dbname_`.

File: db/db_impl_compaction_flush.cc (hunk 2 above)
Comment (ajkr) [reply db/db_impl_compaction_flush.cc]: That's a really good point. We are over-reserving a bit already because the size of the input files is typically larger than the size of the output files. That's because old versions of overwritten/deleted keys can be dropped during compaction. But yeah, estimating the amount of log+flush data during the compaction would be ideal.

Comment (ajkr) [reply db/db_impl_compaction_flush.cc]: Short sleep here is ok with me, as I don't see a way to be notified of space freeing up.

File: env/env_posix.cc [hunk 4]
```diff
@@ -778,6 +779,18 @@ class PosixEnv : public Env {
     return Status::OK();
   }
 
+  virtual Status CheckCompactionSize(const char *path_in_filesystem,
+      uint64_t size_added_by_compact) override {
+    // Figure out how much free space is in the filesystem
+    struct statvfs statbuf;
+    statvfs(path_in_filesystem, &statbuf);
+    uint64_t freespace = statbuf.f_bsize * statbuf.f_bfree;
+    if (freespace <= size_added_by_compact)
```
Comment (ajkr) [inline env/env_posix.cc]: we typically surround single-line conditional bodies with curly braces

File: env/env_posix.cc [hunk 5]
```diff
@@ -778,6 +779,18 @@ class PosixEnv : public Env {
     return Status::OK();
   }
 
+  virtual Status CheckCompactionSize(const char *path_in_filesystem,
```
Comment (ajkr) [inline env/env_posix.cc]: I don't really feel it fits that well with the Env abstraction. Would it make sense to have a `GetEstimatedFreeSize` function here, and put the logic for checking whether compaction can run closer to the compaction code?

File: db/db_impl_compaction_flush.cc [hunk 6]
```diff
@@ -1541,27 +1541,48 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
+          for (size_t j = 0; j < c->num_input_files(i); j++) {
+            FileMetaData *filemeta = c->input(i, j);
+            size_added_by_compaction += filemeta->fd.GetFileSize();
+          }
+        }
+        Status s = env_->CheckCompactionSize(dbname_.c_str(), size_added_by_compaction);
+        if (!s.ok() && !s.IsNotSupported()) {
+          // Then don't do the compaction
+          c->ReleaseCompactionFiles(status);
+          // TODO(amytai09): Is there any other cleanup we need to do?
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: we should probably log something here so people can debug why their compactions stopped

File: env/env_posix.cc [hunk 7]
```diff
@@ -778,6 +779,18 @@ class PosixEnv : public Env {
...
+      uint64_t size_added_by_compact) override {
+    // Figure out how much free space is in the filesystem
+    struct statvfs statbuf;
+    statvfs(path_in_filesystem, &statbuf);
```
Comment (ajkr) [inline env/env_posix.cc]: probably should check the return value

File: db/db_impl_compaction_flush.cc (hunk 6 above)
Comment (amytai) [reply db/db_impl_compaction_flush.cc]: done

File: env/env_posix.cc (hunk 5 above)
Comment (amytai) [reply env/env_posix.cc]: Gone now

File: env/env_posix.cc (hunk 7 above)
Comment (amytai) [reply env/env_posix.cc]: Gone now

File: env/env_posix.cc (hunk 4 above)
Comment (amytai) [reply env/env_posix.cc]: Gone now

**Comment by ajkr** [comment]
seems a lot of unrelated commits made their way into this PR. any idea why?

**Comment by amytai** [comment]
I fetched/merged from upstream / rocksdb master branch.

File: db/db_sst_test.cc [hunk 8]
```diff
@@ -538,6 +538,46 @@ TEST_F(DBSSTTest, DBWithMaxSpaceAllowed) {
...
+
+  // Generate another file to trigger compaction.
+  for (int i = 0; i < 10; i++) {
+    ASSERT_OK(Put(Key(i), RandomString(&rnd, 50)));
+  }
+  ASSERT_OK(Flush());
+  dbfull()->TEST_WaitForCompact();
+
+  ASSERT_GT(cancelled_compaction, 0);
+  rocksdb::SyncPoint::GetInstance()->DisableProcessing();
```
Comment (ajkr) [inline db/db_sst_test.cc]: maybe also good to test that, by the end, the reserved space for compaction is back to zero.

Also have you tested that it can continue after a failure to reserve space? For example, you could call `SetMaxAllowedSpace` in your callback to make the space big enough. (You'd also need to change TEST_WaitForCompact to wait for unscheduled_compactions_ so it doesn't return early like it does now.)

File: db/db_impl_compaction_flush.cc [hunk 9]
```diff
@@ -1618,27 +1621,57 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
+          c->ReleaseCompactionFiles(status);
+
+          ROCKS_LOG_BUFFER(log_buffer, "Cancelled compaction because not enough room");
           AddToCompactionQueue(cfd);
           ++unscheduled_compactions_;
-          MaybeScheduleFlushOrCompaction();
+
+          c.reset();
+          // Sleep for 1 ms?
+          env_->SleepForMicroseconds(1000000);
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: I realized this is also done in BackgroundCallCompaction (with db mutex released). You should be able to get rid of this one and it won't loop too fast, right?

Comment (amytai) [reply db/db_impl_compaction_flush.cc]: Yes, looks like it'll get called because status.ok(). Removed the sleep.

File: db/db_sst_test.cc (hunk 8 above)
Comment (amytai) [reply db/db_sst_test.cc]: Good call, the compaction actually doesn't get rescheduled because the compaction scores are inaccurate.

Comment (ajkr) [reply db/db_sst_test.cc]: It's probably because we only pick compaction if there's a level with score greater than one. So we need to know score first in order to know we should pick compaction.

File: include/rocksdb/sst_file_manager.h [hunk 10]
```diff
@@ -11,6 +11,7 @@
 #include <vector>
 
 #include "rocksdb/status.h"
+#include "db/compaction.h"
```
[the remark this answers is not in the document]
Comment (ajkr) [reply include/rocksdb/sst_file_manager.h]: yup that's fine, only things under "include/" are considered public headers which can't include internal (non-"include/") headers.

File: db/compaction_picker.cc [hunk 11]
```diff
@@ -1083,6 +1083,7 @@ void LevelCompactionBuilder::PickFilesMarkedForCompaction() {
 void LevelCompactionBuilder::SetupInitialFiles() {
   // Find the compactions by size on all levels.
   bool skipped_l0_to_base = false;
+  vstorage_->ComputeCompactionScore(ioptions_, mutable_cf_options_);
```
Comment (ajkr) [inline db/compaction_picker.cc]: I didn't quite get why this works. My understanding is compaction picking works like this:

1. `LevelCompactionPicker::NeedsCompaction` is called and checks if any level has score >= 1. If yes, we push the column family to the compaction queue
2. A background thread dequeues a column family from the compaction queue and calls `PickCompaction` on it. That function chooses which level to compact based on highest score.

I thought recomputing score (if we weren't doing it already) would have to be done before 1. as otherwise the levels might all have scores falsely below 1 (since score was last computed based on the assumption that the now-canceled compaction would successfully complete). But this is doing it in the beginning of 2. so I'm not sure how it helps compaction get rescheduled. Maybe we should debug :).

Comment (amytai) [reply db/compaction_picker.cc]: So if the compaction gets cancelled, the column family is automatically pushed back into the compaction queue in db_impl_compaction_flush.cc, so PickCompaction is called at some point AFTER a compaction has been cancelled. However, PickCompaction was coming up with no compaction, because the compaction score of level0 was 0, because the PREVIOUS compaction picker *thought* it successfully compacted the 2 level0 files we have. Unfortunately, this compaction was cancelled, so this is a stale score.

Anyway, so if don't recompute compaction score here, the restarted, previously cancelled compaction, mistakenly has a stale score of 0.

Comment (ajkr) [reply db/compaction_picker.cc]: Oh ok, 1. is skipped because we use `AddToCompactionQueue` to requeue immediately. Let me think about it more. I suspect we should move it to after the sleep since, if there are multiple background threads, it could be picked up again immediately.

Comment (amytai) [reply db/compaction_picker.cc]: OK, or we could just increment unscheduled_compactions_ in db_impl_compaction_flush.cc to make sure background functions know that a compaction needs to be done. But then, we still need to recompute the compaction score somewhere (perhaps in NeedsCompaction right before it checks CompactScore). Then cfd can be organically requeued. I don't actually know how much this buys us though. What exactly is the goal here? We just want to delay  when a cancelled compaction is rescheduled?

Comment (ajkr) [reply db/compaction_picker.cc]: I didn't have a goal above since I didn't fully understand yet. But now, yes, that is the primary goal. A secondary goal would be to put the call to recompute compaction score in the narrowest place where it's necessary. I agree it's idempotent, but the code is more understandable if we only call it in situations where compaction score could've changed.

I think a good place to put it is in the error handling code in `BackgroundCallCompaction`, after the sleep. It might even be worthwhile to move the `ReleaseCompactionFiles` there too, so other compaction threads don't reattempt the same failing work before the delay finishes. But I'll defer that decision to you.

File: db/db_impl_debug.cc [hunk 12]
```diff
@@ -126,7 +126,7 @@ Status DBImpl::TEST_WaitForCompact() {
 
   InstrumentedMutexLock l(&mutex_);
   while ((bg_bottom_compaction_scheduled_ || bg_compaction_scheduled_ ||
-          bg_flush_scheduled_) &&
+          bg_flush_scheduled_ || (waitUnscheduled && unscheduled_compactions_)) &&
```
Comment (ajkr) [inline db/db_impl_debug.cc]: the camel-case is more for function names. variable names can be underscore-separated lowercase (we follow google style guide if you're interested)

File: db/db_sst_test.cc [hunk 13]
```diff
@@ -538,6 +538,55 @@ TEST_F(DBSSTTest, DBWithMaxSpaceAllowed) {
   ASSERT_NOK(Flush());
 }
 
+TEST_F(DBSSTTest, CancellingCompactionsWorks) {
```
Comment (ajkr) [inline db/db_sst_test.cc]: nice test case!

File: db/compaction_picker.cc (hunk 11 above)
Comment (amytai) [reply db/compaction_picker.cc]: Hrm, it's actually quite difficult to both call ComputeCompactionScore and ReleaseCompactionFiles in the error handling code in BackgroundCallCompaction, because by then the compaction has been reset, so the cfd has been released. We can obviously pass up the cfd, if we encounter CompactionTooLarge status (I modified the status because I realized I created a status and never used it), but that seems inelegant.

So, I put ComputeCompactionScore right next to ReleaseCompactionFiles, because that's where you can access the cfd... If you have better ideas, I am open to them.

Comment (ajkr) [reply db/compaction_picker.cc]: OK that's fine for me. Another solution would be to move the sleep logic into `BackgroundCompaction` and refactor that function to have a single exit point. Then we can access cfd and release/recompute after the sleep. It's kind of tricky so one of us can try it in a follow-up PR if you agree it's reasonable.
